### PR TITLE
[Model] Add Z-Image step execution support

### PR DESCRIPTION
**Pull request by Xenoryn**
## Purpose

Z-Image currently uses a monolithic, request-level `forward()`. Each request has to wait for the previous one to finish all 8 steps.

This PR adds `--step-execution` support, allowing requests to interleave by step and share GPU time within the same tick.

On 1× H800 with 8 steps, FA3, `torch.compile`, and CFG off:

- 512², concurrency=2 vs serial: **+23.4%** QPS (3.79 vs 3.07)
- 1024², concurrency=2 vs serial: **+8.5%** QPS (0.93 vs 0.86)

### What changed

| File | Change |
| --- | --- |
| `pipeline_z_image.py` | Adds the `prepare_encode`, `denoise_step`, `step_scheduler`, and `post_decode` stage hooks. `forward()` now reuses `_prepare_generation_context`, removing about 150 duplicate lines. |
| `z_image_transformer.py` | Adds the variable-length batch support required by step execution. |
| Tests, perf config, and docs | Adds regression coverage, a reproduction script, and updates the feature matrix. |

### Why the transformer changes are required

| Change | Without it |
| --- | --- |
| Wire up the commented-out `attn_mask=` path | **Silently incorrect images:** padding tokens enter softmax when prompts have different lengths, shifting the subject for shorter prompts. |
| Use the `mask=None` fast path for equal-length batches | About 40% lower performance due to the masked SDPA kernel and `torch.compile` graph cache misses. |
| Reject ring attention with variable-length inputs | **Silently incorrect images:** the ring kernel ignores the padding mask and corrupts cross-device QK results. |
| Clear `attn_mask=None` in unified blocks for equal-length inputs | About 50–70% lower performance because the main transformer stack takes the masked path and falls back to SDPA when ring degree > 1. |

The first two are correctness fixes. Step execution may run without them, but its output is wrong. The last two preserve the performance gain.

### Related cleanup

- Moves guidance, CFG, joint-attention, and interrupt values from shared instance fields to per-request `state.extra["z_image_*"]`, preventing cross-request state contamination.
- Removes three dead `if self.interrupt: return` checks that are never toggled under continuous batching.
- Deduplicates about 150 lines of prompt parsing, image extraction, validation, encoding, latent setup, and timestep logic. `forward()` is now a thin wrapper around the shared helper, denoising loop, and decode.

## Test Plan

Run the Z-Image attention-mask and step-execution regression tests:

```bash
pytest -q tests/diffusion/models/z_image/test_zimage_attention_mask.py \
    tests/diffusion/test_diffusion_step_pipeline.py::TestSupportedPipelines \
    tests/diffusion/test_diffusion_step_pipeline.py::TestZImageStepExecution
```

Run the performance harness on 1× H800:

```bash
export DIFFUSION_ATTENTION_BACKEND=FLASH_ATTN
export PERF_CONFIG=dfx/perf/tests/test_zimage_step_execution_vllm_omni.json
bash tests/dfx/perf/scripts/run_zimage_step_execution_pr.sh perf
```

> [!IMPORTANT]
> `benchmark_params` must explicitly include `extra-body: {guidance_scale: 0.0}`. The benchmark client does not set it automatically. Since #4999, the server maps `None → 1.0`, silently enabling CFG and cutting QPS in half.

**vLLM Version:** 0.25.0

**vLLM-Omni Commit:** c1a2046730693f2962f0a362d9f765e1c26af104

### CPU regression tests

Passed locally and on `origin/main`:

```text
10 passed
```

Coverage includes:

- Padding masks do not affect real rows.
- Ring attention rejects variable-length inputs.
- Per-row CFG gating handles zero-scale, all-active, and all-truncated cases.
- Late I2I inputs are promoted to FP32.
- Deferred `z_image_cfg_normalization` values are restored.

### Performance

| Resolution | Serial QPS | Step c=2 QPS | Gain |
| --- | ---: | ---: | ---: |
| 512² | 3.07 | **3.79** | **+23.4%** |
| 1024² | 0.86 | **0.93** | **+8.5%** |

<details>
<summary>Full benchmark results</summary>

| Mode | Resolution | Completed/Requests | QPS |
| --- | ---: | ---: | ---: |
| Serial | 512² | 16/16 | 3.07 |
| Step c=1 | 512² | 16/16 | 2.80 |
| Step c=2 | 512² | 32/32 | **3.79** |
| Step c=2, varlen | 512² | 32/32 | 2.42 |
| Serial | 1024² | 8/8 | 0.86 |
| Step c=1 | 1024² | 8/8 | 0.85 |
| Step c=2 | 1024² | 16/16 | **0.93** |

</details>

### Visual check

Ran two prompts of different lengths with `seed=42`, once at c=1 and once at c=2. The results are not pixel-identical because batch shape changes FP16 accumulation order and advances the shared generator differently.

They are semantically consistent: same subject, composition, and lighting, with no CFG oversaturation, cross-prompt bleed, or padding artifacts.

### CODE REVIEW

**Comment by Xenoryn** [comment]
Hi @asukaqaq-s @yJader, this PR extends diffusion step execution to Z-Image.

Could you help point me to the right reviewers for this area? A review from you would also be very welcome if you have time. Thanks!

File: vllm_omni/diffusion/sched/step_scheduler.py [hunk 1]
```diff
@@ -39,6 +41,18 @@ def _reset_scheduler_state(self) -> None:
...
+        # extra_args, while all current step-execution pipelines otherwise use
+        # 50, so resolve the same value the worker will consume before admission.
+        # Explicit timestep or sigma schedules already define the total and do
+        # not require mutating num_inference_steps.
+        sampling = request.sampling_params
+        if sampling.num_inference_steps is None and sampling.timesteps is None and sampling.sigmas is None:
+            extra_args = sampling.extra_args or {}
+            sampling.num_inference_steps = int(
+                extra_args.get("num_inference_steps", _STEP_EXECUTION_DEFAULT_NUM_INFERENCE_STEPS)
+            )
```
Comment (yJader) [inline vllm_omni/diffusion/sched/step_scheduler.py]: I don't think SamplingParams should use a global default here, since None intentionally means “not specified” and inference-step defaults are model-specific. Could we resolve the default in a model-aware request normalization stage before scheduling, or let StepScheduler handle an unknown total until the worker prepares the timesteps?

File: vllm_omni/diffusion/sched/base_scheduler.py [hunk 2]
```diff
@@ -23,19 +24,45 @@
...
+
+    return tuple(sorted(str(name) for name, value in multi_modal_data.items() if value is not None))
+
+
 def get_sampling_params_key(request: OmniDiffusionRequest) -> SamplingParamsKey:
     """Build a batch-compatibility key from the request's sampling params."""
     sampling = request.sampling_params
     lora_request = getattr(sampling, "lora_request", None)
     return SamplingParamsKey(
+        input_mode=get_request_input_mode_key(request),
```
Comment (yJader) [inline vllm_omni/diffusion/sched/base_scheduler.py]: `SamplingParamsKey` is missing the input_mode field passed here, which causes a TypeError. Please add `input_mode: tuple[str, ...] = ()`

File: vllm_omni/diffusion/attention/parallel/ulysses.py [hunk 3]
```diff
@@ -208,6 +240,7 @@ def pre_attention(
         joint_strategy = "front"
         joint_len = 0
         joint_orig_head_cnt = 0
+        local_query_len = int(query.shape[0])
```
Comment (yJader) [inline vllm_omni/diffusion/attention/parallel/ulysses.py]: `query` has shape `[B, S_local, H, D]`, so `shape[0]` is the batch size rather than the local sequence length. 

```suggestion
        local_query_len = int(query.shape[1])
```

Comment (lishunyang12) [inline vllm_omni/diffusion/attention/parallel/ulysses.py]: `query` here is `(B, S_local, H, D)`, so `local_query_len` is the batch size, but below it's compared against `attn_mask.shape[1]` (the sequence len). This should be `query.shape[1]`, otherwise `mask_len == local_query_len` never fires and real Ulysses batches hit the `elif` raise.

File: vllm_omni/diffusion/models/z_image/pipeline_z_image.py [hunk 4]
```diff
@@ -719,6 +726,507 @@ def forward(self, req: DiffusionRequestBatch) -> DiffusionOutput:
...
+        noise_pred: torch.Tensor,
+        **kwargs: Any,
+    ) -> None:
+        del kwargs
+        if self.interrupt:
+            return
+        state.latents = state.scheduler.step(
+            noise_pred.to(torch.float32), state.current_timestep, state.latents, return_dict=False
+        )[0]
+        assert state.latents.dtype == torch.float32
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/z_image/pipeline_z_image.py]: This `assert` guards the mixed-dtype CB-join invariant set up in `prepare_encode`, but it's stripped under `python -O`. Raise a `ValueError` instead so it holds in prod.

File: vllm_omni/diffusion/models/z_image/pipeline_z_image.py [hunk 5]
```diff
@@ -719,6 +726,507 @@ def forward(self, req: DiffusionRequestBatch) -> DiffusionOutput:
...
+        if input_batch.negative_prompt_embeds is None:
+            negative_prompt_embeds = None
+        else:
+            negative_prompt_embeds = self._padded_prompt_embeds_to_list(
+                input_batch.negative_prompt_embeds,
+                input_batch.negative_prompt_embeds_mask,
+            )
+
+        actual_batch_size = int(input_batch.latents.shape[0])
+        timestep = input_batch.timesteps.to(device=input_batch.latents.device, dtype=torch.float32)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/z_image/pipeline_z_image.py:952]: This scalar/1D broadcast+validate pattern shows up 3x (`timestep`, `guidance_scales`, and inside `_extra_row_tensor`). Pull it into one `_broadcast_rows(t, n, name)` helper and reuse.

**Review by lishunyang12 — COMMENTED** [review]
The ulysses `query.shape[0]` vs `[1]` looks like an axis bug.

File: vllm_omni/diffusion/sched/base_scheduler.py (hunk 2 above)
Comment (Xenoryn) [reply vllm_omni/diffusion/sched/base_scheduler.py]: Thanks for catching this. The `input_mode` field was accidentally ommited and has been restored in 919e99a.

File: vllm_omni/diffusion/attention/parallel/ulysses.py (hunk 3 above)
Comment (Xenoryn) [reply vllm_omni/diffusion/attention/parallel/ulysses.py]: You're right, the local sequence length should be `query.shape[1]`. This was a typo in the submitted code and has been corrected in 919e99a. Thanks.

Comment (Xenoryn) [reply vllm_omni/diffusion/attention/parallel/ulysses.py]: Right, the local sequence length should be `query.shape[1]`. This was a typo in the submitted code and has been corrected in https://github.com/vllm-project/vllm-omni/commit/919e99a717fe62a6f2d38048214c2c2dfd362800. Thanks ~

File: vllm_omni/diffusion/models/z_image/pipeline_z_image.py (hunk 4 above)
Comment (Xenoryn) [reply vllm_omni/diffusion/models/z_image/pipeline_z_image.py]: Agreed. I replaced the assertion with an explicit `ValueError` so the invariant is preserved under `python -O`. Fixed in 2a14889.

File: vllm_omni/diffusion/models/z_image/pipeline_z_image.py (hunk 5 above)
Comment (Xenoryn) [reply vllm_omni/diffusion/models/z_image/pipeline_z_image.py:952]: Good point. I extracted the shared broadcasting logic into `_broadcast_rows` and reused it in all three places. Updated in 2a14889.

File: vllm_omni/diffusion/sched/step_scheduler.py (hunk 1 above)
Comment (Xenoryn) [reply vllm_omni/diffusion/sched/step_scheduler.py]: Thanks for the suggestion. I removed the global default and updated `StepScheduler` to allow an unknown total until the worker prepares the timesteps. Fixed with test coverage in 9b5b7a7.

**Comment by Xenoryn** [comment]
Quick performance note on #4999:

The current numbers were collected while `guidance_scale=0.0` was silently normalized to `1.0`, which unintentionally enabled CFG for Z-Image-Turbo. Once #4999 lands, I’ll explicitly set `guidance_scale=0.0` and refresh the PR body with CFG-disabled results.

A preliminary 3-run validation with the equivalent local fix (1× NVIDIA H800, 8 steps, FLASH_ATTN; 384 requests, 0 failures) gives:

| Resolution | Serial QPS | Step C=2 QPS | Gain |
| --- | ---: | ---: | ---: |
| `512x512` | 3.0006 ± 0.0177 | 3.8026 ± 0.0865 | **+26.7%** |
| `1024x1024` | 0.8626 ± 0.0140 | 0.9269 ± 0.0199 | **+7.5%** |

These will replace the current +10.2% / +12.2% figures, along with an updated environment note.

The feedback raised so far has been addressed. This PR has been open for a little over 3 weeks, so whenever reviewers have a chance, a fresh look—especially at the interaction with #4999—would be greatly appreciated. Thanks!

**Comment by hsliuustc0106** [comment]
do we still need to make so many changes to support step-execution for one model? cc @bjf-frz @asukaqaq-s

**Comment by asukaqaq-s** [comment]
My understanding is that batching support should mainly require changes to `pipeline_Z_IMAGE`, with at most a few additional changes in the Transformers-related files. Why does this PR touch so many files, including parallelism and the scheduler?

Could changes unrelated to step-wise execution be simplified or split into separate PRs?

We are also refactoring pipeline support for stage disaggregation and batching. Please also take a look at #4948.

**Comment by Xenoryn** [comment]
> My understanding is that batching support should mainly require changes to `pipeline_Z_IMAGE`, with at most a few additional changes in the Transformers-related files. Why does this PR touch so many files, including parallelism and the scheduler?
> 
> Could changes unrelated to step-wise execution be simplified or split into separate PRs?
> 
> We are also refactoring pipeline support for stage disaggregation and batching. Please also take a look at #4948.

Hi @asukaqaq-s, thanks for the review — makes sense, splitting it up.

A bit awkward landing such a big diff in one go — got a bit ahead of myself Orz. In fact, some of the shared-code changes in there are useful, so we'll follow up with them separately (the `Ulysses attn_mask` fix you flagged, a small scheduler change to allow pipeline-owned `num_inference_steps`, an interface opt-out flag, …).

So, plan for #4994: trim it down to just the two Z-Image files (`pipeline_z_image.py` + `z_image_transformer.py`, plus `tests/perf/doc`).

One thing we're unsure about: the 4-stage methods this PR overrides (`prepare_encode` / `denoise_step` / …) look like they'll be replaced by `pack_stage_state` / `unpack_stage_state` / `execute` in #4948. So we're wondering — land this now and remap in a small follow-up once #4948 merges, or wait and go straight onto the new API? Either works for us, mostly depends on your ETA for #4948. What do you prefer?
